core/src/test/java12/com/linecorp/armeria/server/JavaHttpClientUpgradeTest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.io.IOException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.server.logging.LoggingService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class JavaHttpClientUpgradeTest {
+
+    static int maxRequestLength = 100000;
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.idleTimeoutMillis(0);
+            sb.requestTimeoutMillis(0);
+            sb.maxRequestLength(maxRequestLength);
+            sb.decorator(LoggingService.newDecorator());
+            sb.service("/echo", (ctx, req) -> {
+                return HttpResponse.from(req.aggregate().thenApply(agg -> {
+                    return HttpResponse.of(ResponseHeaders.of(200), agg.content());
+                }));
+            });
+        }
+    };
+
+    @Test
+    void shouldHandleLargeData() throws Exception {
```

Comment:
Can we also check which protocol is involved?

Reply:
Added in [`a5ea340` (#3913)](https://github.com/line/armeria/pull/3913/commits/a5ea340618afeabddf679f6e027013c7d992a6e7). By the way, I found a bug that a protocol violation error attempted to write the response with HTTP/1 encoder after sending an upgrade response.
The bug was also fixed in a5ea340 together.